philips: support setting gradient

zegl opened the pull request:
API receives a of comma separated hex RGB values.

Added to 4080248U9 / 915005987201 to start with.

This updates https://github.com/Koenkk/zigbee2mqtt/issues/15243

Depends on https://github.com/Koenkk/zigbee-herdsman/pull/622

Koenkk commented on devices/philips.js:
```diff
@@ -1636,9 +1673,16 @@ module.exports = [
+        ],
+        exposes: [
+            exposes.enum('gradient_scene', ea.SET, Object.keys(gradientScenes)),
+            exposes.text('philips_hue_multicolor', ea.SET),
```
- I suggest to rename this to `colors`
- We can use the new list exposes instead: https://github.com/Koenkk/zigbee-herdsman-converters/blob/4c961732b32d34905b9b04738018c7b7b3b1384a/lib/exposes.js#L117

zegl replied:
Thanks for the quick review! I'll take a look and implement your suggested changes. :-)

zegl commented:
Thanks for your feedback, I've now incorporated your suggestions. 👍 

As a future improvement, I was thinking if it was worth exposing some metadata/property like a [min, max] number of supported colors. What would be the preferred way of achieving this? Would adding `withLengthMin`/`withLengthMax` to the `List` class be a good idea?

Koenkk commented on devices/philips.js:
```diff
@@ -1636,9 +1672,16 @@ module.exports = [
+        ],
+        exposes: [
+            exposes.enum('gradient_scene', ea.SET, Object.keys(gradientScenes)),
+            exposes.list('colors', ea.SET, exposes.text('color', ea.SET)),
```
Can you add a description to the `color`? (required format example). Something like:

`exposes.text('color', ea.SET)` -> `exposes.text('color', ea.SET).withDescription('Color in hex e.g. #DEEB34')`

zegl replied:
Yes, will do! I figured out that it probably makes more sense to use X/Y colors instead of RGB, as that's what the device is using, so I'm rewriting this now. I'll ping you once it's ready for another review.

zegl commented:
Changes since last time:

* Switched the API to XY color space, as this is what the device protocol uses
* Managed to get a fromZigbee working, seems to work pretty well!
* Added support for setting between 1 and 9 colors

Please take a look, and let me know if there is anything I should improve before this can be merged!

zegl commented:
The tests are failing, seems to be related to 56d18d1a49c3eaf3740d603ffe9a48465f6dee51.

Koenkk commented:
I think RGB (hex) colors might be more user friendly as it seems to be hard to figure out x/y values.

zegl commented:
Yeah, agreed that RGB is easier. I changed my mind because of consistency reasons, none of the other converters for Philips devices seems to have RGB APIs, seems like they all are using x/y or h/s. So it felt odd to add RGB/hex to the mix, but I can change it back if you want me to.

Koenkk commented:
Lets change it back, this is typically something that will be entered from the frontend.

zegl commented:
Ok, I think we're getting somewhere. I'm starting to get more comfortable with the codebase now, and have made things a lot better now. 😄 

1. I've added `convertGet` for the gradients (depends on https://github.com/Koenkk/zigbee-herdsman/pull/622 to define `0xFC03` / `0x0002` as "state").
2. Changed the API to use RGB HEX. I added ColorRGB.toHex().
3. Moved the parsing to lib/philips.js and added some basic tests for it. I think that this will be really useful down the line if we add support for more features of the gradient payload.

I've been doing end-to-end tests using the latest `master` branch of zigbee2mqtt, and on top of the PR I made to zigbee-heardsman. Everything seems to be working great!

After this is merged (this PR is starting to get a bit too big in my opinion). I'm planning to add this feature to all Hue Gradient devices, and fix up the list of devices a bit. I'll also add a gradient picker to the frontend project.

Thanks for your support so far, big thanks! Hope we can get this merged. :-)

sjorge commented on devices/philips.js:
```diff
@@ -82,6 +83,26 @@ const fzLocal = {
+            cluster: 'manuSpecificPhilips2',
+            type: ['attributeReport', 'readResponse'],
+            convert: (model, msg, publish, options, meta) => {
+                if (msg.data.hasOwnProperty('2')) {
```
Should the '2' here be 'multiColor' ?

zegl replied:
I probably should have commented on it, it's for compatibility until https://github.com/Koenkk/zigbee-herdsman/pull/622 get's merged which maps `2` to `state` (used below).

zegl replied:
Done!

Koenkk commented:
@zegl I did a small refactor, created a new extend for the gradient to reduce code when adding this for more devices (which I also did). 

- Can you do a final check if everything works?
- I'm wondering instead of `colors`, `gradient_colors` is a better name (since `gradient_scene` is similar to this)

zegl commented:
@Koenkk Nice refactor! I made a test and everything is working as it should. :-)

zegl commented:
@Koenkk I think that `colors` is cleaner, but I'll let you decide. I'd assume that we could deprecate `gradient_scene` soon, as `colors` supports everything that `gradient_scene` does and more.

Koenkk commented:
It seems the way Philips advertises the multicolor feature is gradient (e.g. https://www.philips-hue.com/en-us/p/hue-white-and-color-ambiance-festavia-string-lights/046677580148). So I think it would be good to name it similar in z2m.

sjorge commented:
> It seems the way Philips advertises the multicolor feature is gradient (e.g. https://www.philips-hue.com/en-us/p/hue-white-and-color-ambiance-festavia-string-lights/046677580148). So I think it would be good to name it similar in z2m.
> 
> The scene part can stay for people that are too lazy to make their own set of colors 😄

Maybe just `gradient` then and mark `gradient_scene` as deprecated?

Koenkk commented:
@sjorge that also sounds good!

zegl commented:
> Maybe just gradient then and mark gradient_scene as deprecated?

Sounds good, I'll make the change.

Koenkk commented:
Awesome, big thanks!
